source/s3_buffer_pool.c

```diff
@@ -101,151 +142,178 @@ void aws_s3_buffer_pool_destroy(struct aws_s3_buffer_pool *buffer_pool) {
-        (void *)alloc_ptr, size, buffer_pool->current_alloc_usage);
-        buffer_pool->current_alloc_usage += size;
-    }
+    buffer_pool->primary_allocated += buffer_pool->block_size;
 
-    aws_mutex_unlock(&buffer_pool->mutex);
+on_allocated:
+    buffer_pool->primary_reserved -= size;
+    buffer_pool->primary_used += size;
 
-    return (struct aws_s3_pooled_buffer){.ptr = alloc_ptr, .size = size};
+    return alloc_ptr;
 }
 
-struct aws_s3_pooled_buffer aws_s3_buffer_pool_acquire_buffer(struct aws_s3_buffer_pool *buffer_pool, size_t size) {
+struct aws_byte_buf aws_s3_buffer_pool_acquire_buffer(
+    struct aws_s3_buffer_pool *buffer_pool,
+    struct aws_s3_buffer_pool_ticket *ticket) {
     AWS_PRECONDITION(buffer_pool);
+    AWS_PRECONDITION(ticket);
 
-    if (size < buffer_pool->block_size) {
-        return s_primary_acquire(buffer_pool, size);
+    if (ticket->ptr != NULL) {
+        return aws_byte_buf_from_empty_array(ticket->ptr, ticket->size);
     }
 
     uint8_t *alloc_ptr = NULL;
+
     aws_mutex_lock(&buffer_pool->mutex);
-    if (buffer_pool->max_mem_usage == 0 || buffer_pool->current_mem_usage + size <= buffer_pool->max_mem_usage) {
-        buffer_pool->current_mem_usage += size;
-        alloc_ptr = aws_mem_acquire(buffer_pool->base_allocator, size);
-    } else {
-        aws_s3_buffer_pool_trim(buffer_pool);
-        if (buffer_pool->current_mem_usage + size <= buffer_pool->max_mem_usage) {
-            buffer_pool->current_mem_usage += size;
-            alloc_ptr = aws_mem_acquire(buffer_pool->base_allocator, size);
-        }
-    }
-    
 
-    if (alloc_ptr != NULL) {
-        AWS_LOGF_DEBUG(0, "(MemLim) acquired mem %p with size %zu current alloc usage %zu", 
-        (void *)alloc_ptr, size, buffer_pool->current_alloc_usage);
-        buffer_pool->current_alloc_usage += size;
+    if (ticket->size <= buffer_pool->primary_size_cutoff) {
+        alloc_ptr = s_primary_acquire(buffer_pool, ticket->size);
+    } else {
+        alloc_ptr = aws_mem_acquire(buffer_pool->base_allocator, ticket->size);
+        buffer_pool->secondary_reserved -= ticket->size;
+        buffer_pool->secondary_used += ticket->size;
     }
 
+    ++buffer_pool->acquire_count;
     aws_mutex_unlock(&buffer_pool->mutex);
+    ticket->ptr = alloc_ptr;
 
-    return (struct aws_s3_pooled_buffer){.ptr = alloc_ptr, .size = alloc_ptr == NULL ? 0 : size};
+    return aws_byte_buf_from_empty_array(ticket->ptr, ticket->size);
 }
 
-void aws_s3_buffer_pool_release_buffer(
+void aws_s3_buffer_pool_release_ticket(
     struct aws_s3_buffer_pool *buffer_pool,
-    struct aws_s3_pooled_buffer pooled_buffer) {
-    if (buffer_pool == NULL || pooled_buffer.ptr == NULL || pooled_buffer.size == 0) {
+    struct aws_s3_buffer_pool_ticket *ticket) {
+
+    if (buffer_pool == NULL || ticket == NULL || ticket->ptr == NULL) {
+        if (ticket != NULL) {
+            aws_mem_release(buffer_pool->base_allocator, ticket);
```

Comment:
when releasing a ticket that has no ptr, don't we still need to adjust the `buffer_pool->primary/secondary_reserved`?